src/test/java/org/opensearch/knn/integ/DerivedSourceIT.java

```diff
@@ -0,0 +1,1252 @@
+                .mapping(objectFieldTypeMapping)
+                .isNested(true)
+                .docCount(DOCS)
+                .indexIngestor(context -> {}) // noop for reindex
+                .build()
+
+        );
+        testDerivedSourceE2E(indexConfigContexts);
+    }
+
+    /**
+     * Single method for running end to end tests for different index configurations for derived source. In general,
+     * flow of operations are
+     *
+     * @param indexConfigContexts {@link IndexConfigContext}
+     */
+    @SneakyThrows
+    private void testDerivedSourceE2E(List<IndexConfigContext> indexConfigContexts) {
+        // Make sure there are 6
+        assertEquals(6, indexConfigContexts.size());
+
+        // Prepare the indices by creating them and ingesting data into them
+        prepareOriginalIndices(indexConfigContexts);
+
+        // Merging
+        testMerging(indexConfigContexts);
+
+        // Update. Skipping update tests for nested docs for now. Will add in the future.
+        if (indexConfigContexts.get(0).isNested() == false) {
+            testUpdate(indexConfigContexts);
+        }
+
+        // Delete
+        testDelete(indexConfigContexts);
+
+        // Search
+        testSearch(indexConfigContexts);
+
+        // Reindex
+        testReindex(indexConfigContexts);
+    }
+
+    @SneakyThrows
+    private void prepareOriginalIndices(List<IndexConfigContext> indexConfigContexts) {
+        assertEquals(6, indexConfigContexts.size());
+        IndexConfigContext derivedSourceEnabledContext = indexConfigContexts.get(0);
+        IndexConfigContext derivedSourceDisabledContext = indexConfigContexts.get(1);
+        createKnnIndex(derivedSourceEnabledContext.indexName, derivedSourceEnabledContext.settings, derivedSourceEnabledContext.mapping);
+        createKnnIndex(derivedSourceDisabledContext.indexName, derivedSourceDisabledContext.settings, derivedSourceDisabledContext.mapping);
+        derivedSourceEnabledContext.indexIngestor.accept(derivedSourceEnabledContext);
+        derivedSourceDisabledContext.indexIngestor.accept(derivedSourceDisabledContext);
+        refreshAllIndices();
+        assertDocsMatch(
+            derivedSourceDisabledContext.docCount,
+            derivedSourceDisabledContext.indexName,
+            derivedSourceEnabledContext.indexName
+        );
+    }
+
+    @SneakyThrows
+    private void testMerging(List<IndexConfigContext> indexConfigContexts) {
+        IndexConfigContext derivedSourceEnabledContext = indexConfigContexts.get(0);
+        IndexConfigContext derivedSourceDisabledContext = indexConfigContexts.get(1);
+        String originalIndexNameDerivedSourceEnabled = derivedSourceEnabledContext.indexName;
+        String originalIndexNameDerivedSourceDisabled = derivedSourceDisabledContext.indexName;
+        forceMergeKnnIndex(originalIndexNameDerivedSourceEnabled, 10);
+        forceMergeKnnIndex(originalIndexNameDerivedSourceDisabled, 10);
```

Comment:
should go here with 1 segment? with 10 segment and 50 docs I am not even sure if merges are happening in the backend.

Reply:
nvm. This is already taken care later in the code.